trafficshape/bucket.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trafficshape
+
+import (
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/martian/log"
+)
+
+// Bucket is a generic leaky bucket that drains at a configurable interval and
+// fills at user defined rate. The bucket may be used concurrently.
+type Bucket struct {
+	capacity int64 // atomic
+	fill     int64 // atomic
+
+	t      *time.Ticker
+	closec chan struct{}
+}
+
+var (
+	// ErrBucketOverflow is an error that indicates the bucket has been overflown
+	// by the user. This error is only returned iff fill > capacity.
+	ErrBucketOverflow   = errors.New("trafficshape: bucket overflow")
+	errFillClosedBucket = errors.New("trafficshape: fill on closed bucket")
+)
+
+// NewBucket returns a new leaky bucket with capacity that is drained
+// at interval.
+func NewBucket(capacity int64, interval time.Duration) *Bucket {
+	b := &Bucket{
+		capacity: capacity,
+		t:        time.NewTicker(interval),
+		closec:   make(chan struct{}),
+	}
+
+	go b.loop()
+
+	return b
+}
+
+// Capacity returns the capacity of the bucket.
+func (b *Bucket) Capacity() int64 {
+	return atomic.LoadInt64(&b.capacity)
+}
+
+// SetCapacity sets the capacity for the bucket and resets the fill to zero.
+func (b *Bucket) SetCapacity(capacity int64) {
+	log.Debugf("trafficshape: set capacity: %d", capacity)
+
+	atomic.StoreInt64(&b.capacity, capacity)
+	atomic.StoreInt64(&b.fill, 0)
+}
+
+// Close stops the drain loop and marks the bucket as closed.
+func (b *Bucket) Close() error {
+	log.Debugf("trafficshape: closing bucket")
+
+	b.t.Stop()
+	close(b.closec)
+
+	return nil
+}
+
+// FillThrottle calls the fill function with the available capacity remaining
+// (capacity-fill) and fills the bucket with the number of tokens returned by the
+// fill function. If the remaining capacity is <= 0, FillThrottle will wait for
+// the next drain before running the fill function.
+//
+// If the fill function returns an error it will be returned by FillThrottle
```

Comment:
If the fill function returns an error, it will be returned by FillThrottle (added comma)